lib/createGltf.js

```diff
@@ -13,335 +16,902 @@ module.exports = createGltf;
-        var minMax = array.getMinMax(components);
-
-        var type = (components === 3 ? 'VEC3' : 'VEC2');
-        var accessor = {
-            bufferView : vertexBufferViewId,
-            byteOffset : vertexBufferByteOffset,
-            byteStride : 0,
-            componentType : WebGLConstants.FLOAT,
-            count : count,
-            min : minMax.min,
-            max : minMax.max,
-            type : type
-        };
-
-        vertexBufferByteOffset += buffer.length;
-        vertexBuffers.push(buffer);
-        var accessorId = 'accessor_' + accessorCount++;
-        gltf.accessors[accessorId] = accessor;
-        return accessorId;
-    }
-
-    function addIndexArray(array, uint32Indices) {
-        var buffer = uint32Indices ? array.toUint32Buffer() : array.toUint16Buffer();
-        var componentType = uint32Indices ? WebGLConstants.UNSIGNED_INT : WebGLConstants.UNSIGNED_SHORT;
-        var length = array.length;
-        var minMax = array.getMinMax(1);
-        var accessor = {
-            bufferView : indexBufferViewId,
-            byteOffset : indexBufferByteOffset,
-            byteStride : 0,
-            componentType : componentType,
-            count : length,
-            min : minMax.min,
-            max : minMax.max,
-            type : 'SCALAR'
-        };
-
-        indexBufferByteOffset += buffer.length;
-        indexBuffers.push(buffer);
-
-        var accessorId = 'accessor_' + accessorCount++;
-        gltf.accessors[accessorId] = accessor;
-        return accessorId;
-    }
-
-    function requiresUint32Indices(nodes) {
-        var nodesLength = nodes.length;
-        for (var i = 0; i < nodesLength; ++i) {
-            var meshes = nodes[i].meshes;
-            var meshesLength = meshes.length;
-            for (var j = 0; j < meshesLength; ++j) {
-                // Reserve the 65535 index for primitive restart
-                var vertexCount = meshes[j].positions.length / 3;
-                if (vertexCount > 65534) {
-                    return true;
-                }
-            }
+function getImageChannel(image, index, targetWidth, targetHeight) {
+    var pixels = image.decoded; // RGBA
+    var width = image.width;
+    var height = image.height;
+    var pixelsLength = width * height;
+    var channel = Buffer.alloc(pixelsLength);
+    for (var i = 0; i < pixelsLength; ++i) {
+        var value = pixels.readUInt8(i * 4 + index);
+        channel.writeUInt8(value, i);
+    }
+    if (width !== targetWidth || height !== targetHeight) {
+        channel = resizeChannel(channel, width, height, targetWidth, targetHeight);
+    }
+    return channel;
+}
+
+function writeChannel(pixels, channel, index, width, height) {
+    var pixelsLength = width * height;
+    for (var i = 0; i < pixelsLength; ++i) {
+        var value = channel.readUInt8(i);
+        pixels.writeUInt8(value, i * 4 + index);
+    }
+}
+
+function getMinimumDimensions(images, options) {
```

Comment:
I believe in most cases the artist will be working with same-sized textures, at least for metallic and roughness. Occlusion may be smaller, but at least it can be packed separately. I think the downscale behavior is okay since I wouldn't expect this warning to get hit much (I may be proven wrong though).